src/predictor/gpu_predictor.cu

```diff
@@ -527,7 +527,7 @@ class GPUPredictor : public xgboost::Predictor {
       Predictor::Predictor{generic_param} {}
 
   ~GPUPredictor() override {
-    if (generic_param_->gpu_id >= 0) {
+    if (generic_param_->gpu_id >= 0 && generic_param_->gpu_id < common::AllVisibleGPUs()) {
```

Comment:
I am not happy about this and I am open to suggestions. This happens when the check in learner.cc fails and the gpu_id value stays invalid. The line below than causes the whole JVM to halt.
An alternative solution would be to add another flag to generic_param or Predictor named like `gpu_id_is_valid`, but that seems even more cumbersome.
Also I do not understand why we need to `cudaSetDevice` in destructor.

Reply:
I think `thrust` might try to do some other operations during destruction at the time of writing these weird code.  Without a correct global gpu id context something broke.

Reply:
Will look into this line of code later, see if there's way to remove them.

Reply:
@trivialfis if there is any research/experimentation I can do to help, please let me know.